From 93514862a5909d54ac21d67a36f087644ead1400 Mon Sep 17 00:00:00 2001
From: Tommy Webb <tommy@calyxinstitute.org>
Date: Wed, 5 Apr 2023 18:44:38 -0400
Subject: [PATCH] Mirror TRUST_RESTRICT_USB setting to sysprop

Update the persist.init.trust_restrict_usb property based on the value
of LineageSettings.Global.TRUST_RESTRICT_USB. This allows vendor init
files to take the necessary steps to fully reactivate USB when the
setting is "Always allow new USB gadgets" (0), without requiring the
user to unplug and re-plug a connected device. In particular, this
eases debugging earlier in the boot process, when applicable.

Issue: calyxos#1414
Requires: Ie210055790baeb7234b167d7b26b6adff8f6361a
Change-Id: Ic6c4b3454f0bd0c0ac1942a30fd03732262a9ad2
---
 .../android/server/usb/UsbDeviceManager.java  | 32 +++++++++++++++++--
 1 file changed, 30 insertions(+), 2 deletions(-)

diff --git a/services/usb/java/com/android/server/usb/UsbDeviceManager.java b/services/usb/java/com/android/server/usb/UsbDeviceManager.java
index f20a9d33904f..a31386b339bd 100644
--- a/services/usb/java/com/android/server/usb/UsbDeviceManager.java
+++ b/services/usb/java/com/android/server/usb/UsbDeviceManager.java
@@ -471,12 +471,15 @@ public class UsbDeviceManager implements ActivityTaskManagerInternal.ScreenObser
 
         sEventLogger = new EventLogger(DUMPSYS_LOG_BUFFER, "UsbDeviceManager activity");
 
+        mHandler.updateTrustRestrictUsbProperty();
+
         mContentResolver.registerContentObserver(
                 LineageSettings.Global.getUriFor(LineageSettings.Global.TRUST_RESTRICT_USB),
                 false,
                 new ContentObserver(null) {
                     @Override
                     public void onChange(boolean selfChange) {
+                        mHandler.updateTrustRestrictUsbProperty();
                         mHandler.setTrustRestrictUsb();
                     }
                 }
@@ -673,6 +676,14 @@ public class UsbDeviceManager implements ActivityTaskManagerInternal.ScreenObser
         protected static final String MTP_PACKAGE_NAME = "com.android.mtp";
         protected static final String MTP_SERVICE_CLASS_NAME = "com.android.mtp.MtpService";
 
+        /**
+         * The persistent property which stores the current Restrict USB setting. One-way mirrored
+         * from LineageSettings so that vendor init files can fully reactivate USB connectivity
+         * when Restrict USB is turned off, without requiring the user to unplug and re-plug.
+         */
+        protected static final String TRUST_RESTRICT_USB_PROPERTY =
+                "persist.init.trust_restrict_usb";
+
         private boolean mIsMtpServiceBound = false;
 
         /**
@@ -1902,9 +1913,26 @@ public class UsbDeviceManager implements ActivityTaskManagerInternal.ScreenObser
 
         public abstract void resetCb(int status);
 
-        public void setTrustRestrictUsb() {
-            final int restrictUsb = LineageSettings.Global.getInt(mContentResolver,
+        public int getTrustRestrictUsb() {
+            return LineageSettings.Global.getInt(mContentResolver,
                     LineageSettings.Global.TRUST_RESTRICT_USB, 1);
+        }
+
+        public void updateTrustRestrictUsbProperty() {
+            final int restrictUsb = getTrustRestrictUsb();
+            try {
+                // Vendor init files should react to this change right away when restrictUsb is 0.
+                setSystemProperty(TRUST_RESTRICT_USB_PROPERTY, Integer.toString(restrictUsb));
+                if (DEBUG) Slog.d(TAG, "Successfully updated " + TRUST_RESTRICT_USB_PROPERTY
+                        + " to " + restrictUsb);
+            } catch (Exception e) {
+                Slog.e(TAG, "Failed to set " + TRUST_RESTRICT_USB_PROPERTY
+                        + " to " + restrictUsb, e);
+            }
+        }
+
+        public void setTrustRestrictUsb() {
+            final int restrictUsb = getTrustRestrictUsb();
             // Effective immediately, ejects any connected USB devices.
             // If the restriction is set to "allow when unlocked", only execute once USB is
             // disconnected and keyguard is showing, to avoid ejecting connected devices on lock,
-- 
GitLab

